src/App.js

```diff
@@ -23,49 +25,107 @@ const generateSquares = () => {
   }
 
   return squares;
-}
+};
 
 const App = () => {
-
   // This starts state off as a 2D array of JS objects with
   // empty value and unique ids.
+
   const [squares, setSquares] = useState(generateSquares());
+  const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1);
+  const [winner, setWinner] = useState(null);
+
+  const onClickCallback = (id) => {
+    setSquares((squares) =>{
+      let newBoard = squares.map((square)=>{
+        for (let property of square){
+          if (property.id === id && property.value === ''){
+            if (currentPlayer === PLAYER_1){
+              property.value = PLAYER_1;
+            } else if (currentPlayer === PLAYER_2){
+              property.value = PLAYER_2;
+            }
+          }
+        }
+        return square;
+      });
+    setWinner(checkForWinner());
+    
+    return newBoard;
+    });
+
+    if (currentPlayer === PLAYER_1){
+      setCurrentPlayer(PLAYER_2);
+    }else {
+      setCurrentPlayer(PLAYER_1);
+    }
+  }; 
 
   // Wave 2
-  // You will need to create a method to change the square 
+  // You will need to create a method to change the square
   //   When it is clicked on.
   //   Then pass it into the squares as a callback
 
-
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if 
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if 
-    //    all three squares have the same value.
+    let i = 0;
+    // Check all the rows and columns for a winner
+    while (i < 3) {
+      if (
+        squares[i][0].value === squares[i][1].value &&
+        squares[i][2].value === squares[i][1].value &&
+        squares[i][0].value !== ''
+      ) {
+        return squares[i][0].value;
+      } else if (
+        squares[0][i].value === squares[1][i].value &&
+        squares[2][i].value === squares[1][i].value &&
+        squares[0][i].value !== ''
+      ) {
+        return squares[0][i].value;
+      }
+      i += 1;
+    }
+    // Check Top-Left to bottom-right diagonal
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][0].value;
+    }
 
-  }
+    // Check Top-right to bottom-left diagonal
+    if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][2].value;
+    }
+
+    return null;
+  };
 
   const resetGame = () => {
     // Complete in Wave 4
-  }
+    setSquares(generateSquares());
+    setWinner(null); 
+    setCurrentPlayer(PLAYER_1);
+  };
+
```

Comment:
Yes. Resetting the game basically means setting all the state back to their initial values.